### ISSUE TITLE
Invalid used by url on storage pool

### ISSUE THREAD

**Issue by edlerd** [canonical/lxd#12820]
# Required information

 * Distribution: Ubuntu snap
 * Distribution version: 5.20-c4bbef1
 * Storage backend in use: Ceph

# Issue description

When using a ceph storage pool and LXD finds orphaned images, they are reported in the used by field of the storage pool api response with an invalid url.

I encountered locally three such images, which were part of the `metadata.used_by` field of the `GET 1.0/storage-pools/my-ceph-pool?recursion=1` response. The urls were of the format `"/1.0/storage-pools/my-ceph-pool/volumes/image/93de62f5c2a8cee537636264e33173f2bbb8fe219eb1a83253119ad2bb37cef8`. When querying this url, I got a 404, hinting those urls are invalid.

Usually, images are reported as `/1.0/images/56e10b38e59e94fd87d16fb6e090747602edd5ee77d9d8ffa6db4efa5821e0a9`.
I suggest the URL should either match the format with the `1.0/images` prefix, or be filtered from the response as the image was removed anyhow.

**Comment by tomponline** [comment]
> Usually, images are reported as `/1.0/images/56e10b38e59e94fd87d16fb6e090747602edd5ee77d9d8ffa6db4efa5821e0a9`.
> I suggest the URL should either match the format with the `1.0/images` prefix, or be filtered from the response as the image was removed anyhow.

I suspect the latter option is more appropriate here, as it wouldnt be correct to link to a non-existent image record, and that is the reason they are showing up as volumes in the first place.

The other option is to acknowledge this can occur and make it so the image volume record is retrievable as a volume in its own right at the resource URL indicated rather than returning a 404.

**Comment by gabrielmougard** [comment]
@edlerd URLs of the form `/1.0/storage-pools/my-ceph-pool/volumes/image/93de62f5c2a8cee537636264e33173f2bbb8fe219eb1a83253119ad2bb37cef` are actually valid (see here: https://github.com/canonical/lxd/blob/4e8c581c0df76b170a5e8e36a2c165a1735a40fe/lxd/storage_volumes.go#L1663C1-L1708C74) and I'm not facing a 404 issue for ZFS backed image volumes for an orphaned image on my side. I'm testing with Ceph as of the time of writing this message and will come back with updates. It's just that this 404 error seemed weird and I wanted to test it with ZFS first. But a significant part of the code has been updated 3 weeks ago (see here: https://github.com/canonical/lxd/pull/12754), so in the meantime, is it possible for you to test this again with the `latest/edge` lxd snap ?

**Comment by edlerd** [comment]
I can't reproduce the situation with an orphaned image volume any more. Unfortunately, I removed the storage pool from the issue description, so failing to reproduce the situation right now. Do you have a hint how I could reproduce orphaned images in my local LXD to test this?

**Comment by gabrielmougard** [comment]
Just for terminology, does 'orphaned' image means that the image is stored in a volume but not currently used by an instance? Or is there something else that I'm missing and that is specific to ceph?

**Comment by edlerd** [comment]
> Just for terminology, does 'orphaned' image means that the image is stored in a volume but not currently used by an instance? Or is there something else that I'm missing and that is specific to ceph?

The case I was encountering, when filing the issue, was that there was a volume under `1.0/storage-pools/:pool/volumes` with type image, but I couldn't find an image for that volume under `1.0/images/:volume-id`. That's what I meant with orphaned image, and I don't know how this situation of an image volume without an image record came about to reproduce it today.

**Comment by gabrielmougard** [comment]
Ah I see! I'd naively think about a data corruption issue (maybe an admin removed the underlying image manually inside ceph). Ok, I'll first try to reproduce the 'lack of underlying image' with a local ZFS storage while keeping its volume and check the API call again. @tomponline do you know if we can easily create an 'orphaned' image volume through the `zfs` CLI tool ?

**Comment by gabrielmougard** [comment]
@edlerd are you able to reproduce the above scenario ?

**Comment by edlerd** [comment]
@gabrielmougard Thank you for the detailed explanation. I followed the steps to have an image that is removed from zfs, but still in lxd. The API works in that case as expected. It doesn't reproduce the issue from the ticket description. Tested with lxd 5.20-f3dd836.

I think the initial issue was a bit different, though. Back then I had an entry in the `used_by` field for a storage pool pointing to `/1.0/storage-pools/:pool/volumes/image/:id`, where now I get the expected `/1.0/images/:id`. This might be fixed due to a recent change since the 5.20-c4bbef1 version of lxd in use when the ticket got created, or the root cause was actually something different. I can't test to reproduce with the 5.20-c4bbef1 version, as that is not available any more.
